File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function (n, edges, source, destination) {
    const generateAdjList = function (map) {
        for (const [x, y] of edges) {
            if (!map[x]) {
                map[x] = [];
            }
            if (!map[y]) {
                map[y] = []
            }

            map[x].push(y);
            map[y].push(x);
        }
    }

    const dfs = function(curr) {
        if(curr === destination) return true;
        visited.add(curr);

        for(let neighbor of adjList[curr]) {
            if(visited.has(neighbor)) continue;

            if(dfs(neighbor)) return true;
        }

        return false;
    }

    const visited = new Set();
    const adjList = new Map();
    generateAdjList(adjList);

    return dfs(source);
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
from collections import deque

class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        def generate_adjency_list(adj_list):
            for x, y in edges:
                if x not in adj_list:
                    adj_list[x] = []
                
                if y not in adj_list:
                    adj_list[y] = []

                adj_list[x].append(y)
                adj_list[y].append(x)
        
        def dfs(curr):
            if curr == destination:
                return True

            visited.add(curr)

            for neighbor in adj_list[curr]:
                if neighbor in visited:
                    continue
                
                if(dfs(neighbor)):
                    return True
            
            return False

        adj_list = {}
        generate_adjency_list(adj_list)
        visited = set()

        return dfs(source)
